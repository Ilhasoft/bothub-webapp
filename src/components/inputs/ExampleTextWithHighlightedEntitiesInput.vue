<template>
  <div class="field">
    <div class="example-txt-w-highlighted-entities">
      <div class="example-txt-w-highlighted-entities__input-wrapper">
        <div class="field">
          <div class="control has-icons-right">
            <div
              v-for="(entity, i) in entitiesBlocks"
              :key="i"
              :class="entityClassAttr">
              <span
                :class="[
                  'example-txt-w-highlighted-entities__entity__before',
              ]">{{ entity.before }}</span><span
                :class="[
                  entity.colorClass,
                  'example-txt-w-highlighted-entities__entity__text'
              ]">{{ entity.text }}</span>
            </div>
            <self-adjust-input
              ref="input"
              :placeholder="$t('webapp.trainings.add_a_sentence')"
              v-model="val"
              transparent
              @select="emitTextSelected"
              @click.stop.prevent="emitTextSelected"
              @keyup.stop.prevent="emitTextSelected"
              @keyup.enter="submit()">
              <span
                v-if="hasAppend"
                slot="append">
                <slot name="append" />
              </span>
            </self-adjust-input>
          </div>
        </div>
      </div>
    </div>
  </div>
</template>

<script>
import SelfAdjustInput from '@/components/inputs/SelfAdjustInput';
import { getEntityColor } from '@/utils/entitiesColors';

const components = {
  SelfAdjustInput,
};

export default {
  name: 'ExampleTextWithHighlightedEntitiesInput',
  components,
  props: {
    value: {
      type: String,
      default: '',
    },
    entities: {
      type: Array,
      default: () => ([]),
    },
  },
  data() {
    return {
      val: this.value,
      selectionStart: 0,
      selectionEnd: 0,
    };
  },
  computed: {
    hasAppend() {
      return this.$slots.append;
    },
    entityClassAttr() {
      const classes = ['example-txt-w-highlighted-entities__entity'];

      if (this.size) {
        classes.push(`example-txt-w-highlighted-entities__entity--${this.size}`);
      }

      if (this.hasAppend) {
        classes.push('example-txt-w-highlighted-entities__entity__with-append');
      }

      return classes;
    },
    entitiesBlocks() {
      return this.entities
        .map(({ start, end, entity }) => {
          const color = getEntityColor(
            entity,
          );
          const colorClass = `entity-${color}`;
          const before = this.value.substring(0, start);
          const text = this.value.substring(start, end);
          return {
            start,
            end,
            colorClass,
            before,
            text,
          };
        }).concat([{
          uuid: 'selected',
          start: this.selectionStart,
          end: this.selectionEnd,
          colorClass: 'entity-selected',
          before: this.value.substring(0, this.selectionStart),
          text: this.value.substring(this.selectionStart, this.selectionEnd),
        }]);
    },
  },
  watch: {
    val() {
      this.$emit('input', this.val);
    },
    value() {
      this.val = this.value;
    },
  },
  methods: {
    submit() {
      this.$emit('submit');
    },
    emitTextSelected(event) {
      const { value, selectionStart, selectionEnd } = event.target;
      const selected = value.slice(selectionStart, selectionEnd);

      const startPadding = selected.search(/\S|$/);
      const endPadding = selected.length - selected.trim().length - startPadding;

      this.selectionStart = selectionStart + startPadding;
      this.selectionEnd = selectionEnd - endPadding;

      this.$emit(
        'textSelected',
        this.selectionStart === this.selectionEnd
          ? null
          : { start: this.selectionStart, end: this.selectionEnd },
      );
    },
    clearSelected() {
      this.$nextTick(() => {
        this.$refs.input.deselect();
        this.$refs.input.inputAction('blur');
      });
    },
  },
};
</script>

<style lang="scss" scoped>
@import '~@/assets/scss/variables.scss';

.example-txt-w-highlighted-entities {

  &__append {
    pointer-events: all !important;
    width: $form-append-width !important;
  }

  &__input-wrapper {
    position: relative;
    background-color: white;
  }

  &__input {
    position: inherit;
    z-index: 1;
  }

  &__entity {
    position: absolute;
    top: 0;
    left: 0;
    z-index: 0;
    height: max-content !important;
    word-wrap: break-word;
    padding: 0.3rem 0.9rem;
    font: inherit;
    line-height: 1.5;
    max-width: 100%;
    background: none;
    border: 1px solid rgba(0, 0, 0, 0);
    pointer-events: none;

    &__with-append {
      padding: 0.3rem 4rem 0.3rem 0.9rem;
    }

    &__before,
    &__text {
      color: rgba(0, 0, 0, 0);
      max-width: 100%;
      white-space: pre-line;
      pointer-events: none;
    }

    &__text {
      border-radius: 4px;
      opacity: .5;
    }
  }
}
</style>
