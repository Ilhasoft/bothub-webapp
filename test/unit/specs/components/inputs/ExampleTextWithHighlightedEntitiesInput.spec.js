import BH from 'bh';
import { shallowMount, createLocalVue } from '@vue/test-utils';
import ExampleTextWithHighlightedEntitiesInput from '@/components/inputs/ExampleTextWithHighlightedEntitiesInput';


const localVue = createLocalVue();
localVue.use(BH);

describe('ExampleTextWithHighlightedEntitiesInput.vue', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallowMount(ExampleTextWithHighlightedEntitiesInput, {
      localVue,
      mocks: {
        $t: () => 'some specific text',
      },
    });
  });

  test('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('set text', () => {
    const textareaValue = 'hi kids';
    beforeEach(() => {
      const input = wrapper.find({ ref: 'input' });
      input.element.value = textareaValue;
      input.trigger('input');
    });

    test(`val is ${textareaValue}`, () => {
      expect(wrapper.vm.val).toBe(textareaValue);
    });

    describe('select text', () => {
      beforeEach(() => {
        const input = wrapper.find({ ref: 'input' });
        input.element.setSelectionRange(0, 2);
      });

      test('renders correctly', () => {
        expect(wrapper).toMatchSnapshot();
      });

      test('textSelected event emitted', () => {
        expect(wrapper.emitted('textSelected')).toBeDefined();
      });
    });

    describe('with entity', () => {
      beforeEach(() => {
        wrapper.setProps({ entities: [{ start: 3, end: 7, entity: 'kids' }] });
      });

      test('renders correctly', () => {
        expect(wrapper).toMatchSnapshot();
      });
    });
  });
});
