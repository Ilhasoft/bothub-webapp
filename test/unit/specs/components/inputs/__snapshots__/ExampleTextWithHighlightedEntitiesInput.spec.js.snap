// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExampleTextWithHighlightedEntitiesInput.vue renders correctly 1`] = `
<div class="field">
  <div class="example-txt-w-highlighted-entities">
    <div class="example-txt-w-highlighted-entities__input-wrapper">
      <div class="field">
        <div class="control has-icons-right is-family-primary">
          <div class="example-txt-w-highlighted-entities__entity"><span class="example-txt-w-highlighted-entities__entity__before"></span><span class="entity-selected example-txt-w-highlighted-entities__entity__text"></span></div>
          <self-adjust-input-stub value="" transparent="true" size="normal">
            <!---->
          </self-adjust-input-stub>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ExampleTextWithHighlightedEntitiesInput.vue set text with entity renders correctly 1`] = `
<div class="field">
  <div class="example-txt-w-highlighted-entities">
    <div class="example-txt-w-highlighted-entities__input-wrapper">
      <div class="field">
        <div class="control has-icons-right is-family-primary">
          <div class="example-txt-w-highlighted-entities__entity"><span class="example-txt-w-highlighted-entities__entity__before"></span><span class="entity-mediterranean-sea example-txt-w-highlighted-entities__entity__text"></span></div>
          <div class="example-txt-w-highlighted-entities__entity"><span class="example-txt-w-highlighted-entities__entity__before"></span><span class="entity-selected example-txt-w-highlighted-entities__entity__text"></span></div>
          <self-adjust-input-stub value="" transparent="true" size="normal">
            <!---->
          </self-adjust-input-stub>
        </div>
      </div>
    </div>
  </div>
</div>
`;
